Editor/HierarchyFolderIcon.cs

```diff
@@ -0,0 +1,161 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Runtime.Remoting.Messaging;
+using System.Xml;
+using UnityEditor;
+using UnityEditor.IMGUI.Controls;
+using UnityEditor.SceneManagement;
+using UnityEditorInternal;
+using UnityEngine;
+using UnityEngine.Experimental.Rendering;
+using UnityHierarchyFolders.Runtime;
+using Object = UnityEngine.Object;
+
+namespace Plugins.UnityHierarchyFolders
+{
+    public class HierarchyFolderIcon
+    {
+        private static Texture2D _openFolderTexture;
+        private static Texture2D _closedFolderTexture;
+        private static Texture2D _openFolderSelectedTexture;
+        private static Texture2D _closedFolderSelectedTexture;
+        
+        private static bool _isInitialized;
+        private static bool _hasProcessedFrame; 
+        
+        // Reflected members
+        private static PropertyInfo _sceneHierarchyProperty;
+        private static PropertyInfo _treeViewProperty;
+        private static PropertyInfo _dataProperty;
+        private static MethodInfo _getRowsMethod;
+        private static MethodInfo _isExpandedMethod;
+        private static PropertyInfo _selectedIconProperty;
+        private static PropertyInfo _objectPPTRProperty;
+        private static MethodInfo _getAllSceneHierarchyWindowsMethod;
+
+
+        [InitializeOnLoadMethod] 
+        static void Startup()
+        { 
+            Initialize();
+            EditorApplication.update += RefreshFolderIcons;
+            EditorApplication.hierarchyWindowItemOnGUI += RefreshFolderIcons;
+        }
+
+        private static Texture2D GetTintedTexture(Texture2D original)
+        {
+            var tinted = new Texture2D(original.width, original.height, 
+            original.graphicsFormat, original.mipmapCount, TextureCreationFlags.MipChain);
+            
+            Graphics.CopyTexture(original, tinted); 
+            var data = tinted.GetRawTextureData<Color32>();
+            for (int index = 0, len = data.Length; index < len; index++)
+            {
+                Color32 c = data[index];
+                byte a = c.a;
+
+                Color.RGBToHSV(c, out float h, out float s, out float v);
+                c = Color.HSVToRGB(0, 0, 1);
+                c.a = a;
+ 
+                data[index] = c;
+            }
+
+            var aaa = tinted.width * tinted.height * 4;
+            var offset = 0; 
+            for (int index = 0; index < tinted.mipmapCount; index++)
+            {
+                tinted.SetPixelData(data, index, offset);
+                aaa >>= 2;
+                offset += aaa;
```

Comment:
offset represents the starting index offset of the current mipmap on the texture.